### ISSUE TITLE
ProtocolWrapper masks the name of the wrapped protocol in the logs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5062]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#5062|
|Type|release blocker: regression|
|Created|2011-04-22 13:37:25Z|
|Branch|https://github.com/twisted/twisted/tree/wrapped-log-prefix-5062-3|

When t.p.policies.ProtocolWrapper is used, it hides the name of the original protocol class in the logs, making it rather useless to have the name in it. It's becoming fairly visibile with the new SSL code, as TLSMemoryBIOProtocol is a ProtocolWrapper subclass.

We should introduce a new way of logging the protocol name, use it in place of `protocol.__class__.__name__` in tcp.py and in place of the `qual(protocol.__class__)` calls in unix.py and udp.py, and customize it in ProtocolWrapper.

Attachments:

* [5062.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8de/8dee13e80a0a63e27f89a3ab22320af4cc0e4290/eacb9a19dea1813057395a4c7ff1e2586e6b2d02.patch) (4809 bytes) - added by retenodus on 2011-07-21 22:16:01Z - Deletion of getattr in LogPrefix, and log prefix lookup and fallback logic shared between TCP, UNIX, and UDP
* [5062-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8de/8dee13e80a0a63e27f89a3ab22320af4cc0e4290/932e5534a3ea6dc5a1a72d08e1cd2571e5db3f4e.patch) (5665 bytes) - added by retenodus on 2011-07-22 13:01:35Z - This patches is the same of previous patch except there are unittest for LogPrefix. I think it could replaces tests of logprefix for tcp, udp and unix, isn't it ?

<details><summary>Searchable metadata</summary>

```
trac-id__5062 5062
type__release_blocker__regression release blocker: regression
reporter__therve therve
priority__normal normal
milestone__Twisted_11_1 Twisted-11.1
branch__branches_wrapped_log_prefix_5062_3 branches/wrapped-log-prefix-5062-3
branch_author__exarkun__therve exarkun, therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1303479445000000 1303479445000000
changetime__1317507835000000 1317507835000000
version__None None
owner__exarkun exarkun
cc__retenodus
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

See [#5073](https://github.com/twisted/twisted/issues/5073) for a similar (not duplicate) issue.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [31750]) Branching to 'wrapped-log-prefix-5062'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

This is ready to review. I'm not sure about a couple of things, (like making IProtocol.logPrefix an attribute), but otherwise I think it's mostly good.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_6"></a>|@zseil set owner to @therve|
|-|-|

Hello,

Replying to [therve](#note_5):
> This is ready to review. I'm not sure about a couple of things, (like making IProtocol.logPrefix an attribute), but otherwise I think it's mostly good.

Thanks for working on this! Here are my comments:

1. The parenthesis in `ProtocolWrapper.logPrefix` seems misaligned. Is it needed for pretty log output?

2. `ProtocolWrapper.logPrefix` should also fallback to class name if the `wrappedProtocol` doesn't have a `logPrefix` attribute.

3. Having `logPrefix` as an attribute of the `Protocol` instead of a method breaks `ProtocolWrapper`'s implementation of `ILoggingContext`, which is provided by most transports. I.e. `callWithLogger(self.transport, ...)` worked previously, but doesn't after your changes.

4. I would prefer if the new attribute (or a method, if you decide for that) wouldn't be added to the `IProtocol` interface, but be part of a separate interface instead (maybe you can just reuse ILoggingContext for that). `Protocol` (or even better, just `ProtocolWrapper`) can than implement that interface, and the transport `logPrefix` methods can then check if `interface.providedBy(self.protocol)` to decide if they should to use the class name fallback. This would prevent breaking third party implementations of `IProtocol` that don't inherit from `Protocol` and `Protocol` subclasses that already use `logPrefix` attribute for some other purpose.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

> This would prevent breaking third party implementations of IProtocol

This seems particularly important.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Thanks, everything handled I think.

Note that I wasn't really worried about breaking IProtocol because interfaces are mainly a declarative tool. Until you call verifyObject, nothing is going to break (and it's not really meant to be called outside of a test suite). The change is removed anyway.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.

  1. Since `Protocol` implements `ILoggingContext` now, wouldn't it be better to use `ILoggingContext.providedBy` instead of `getattr(proto, "logPrefix", ...)`?
  1. Can the log prefix lookup and fallback logic be shared between TCP, UNIX, and UDP?  (I think that `twisted.internet.test.connectionmixins.ConnectionTestsMixin` might be a good approach for this).
  1. Can the tests for at least TCP and UNIX share some code?  Endpoints might help with this.
  1. IOCP is still broken.  The new unit tests do manage to demonstrate this, as long as you run 'trial --reactor=iocp' on Windows.  However, since I would like to stop doing this someday...

What is the barrier to having new tests written in `twisted.internet.test` instead of `twisted.test` and based on `ReactorBuilder` instead of using the global reactor?  It seems to me that the ideal is that we would write all new reactor-using tests using `ReactorBuilder` and port all such old tests as well.

Alternatively, do these tests need a reactor at all?  They could be whitebox tests that instantiate a `BaseClient` and call a method on it, like `_determineLogPrefix`, and just assert something about that method's return value.  If the method were on a base class high enough in the hierarchy, maybe the same tests would even suffice to cover both `PosixBaseReactor` and `IOCPReactor`.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|retenodus set owner to retenodus<br>retenodus set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|retenodus commented|
|-|-|

I don't understand what do you mean in the point 2 : do you think we should have a class LogPrefix which would be used by TCP, UNIX and UDP, or am I misunderstanding ? And I don't understand in what ConnectionTestsMixin is related with this.

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|retenodus removed owner<br>retenodus set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @therve|
|-|-|

Hi retenodus,

Thanks for the contribution, but I don't think that this patch really addresses the feedback from the previous review.  Here is some review feedback which may be useful for your future contributions, however:

  1. The new `logpre` module is a public API I definitely wouldn't want to support, and it's mostly unnecessary and redundant, since it just does the same check that was done previously.
  2. A minor stylistic note: I'm not sure if the coding standard actually says this anywhere, but you should not put a space before the colon in 'try:', 'except:', 'lambda:', etc.
  3. The log prefix tests would need docstrings, explaining what they're testing.
  4. There would need to be a copyright notice and docstring at the top of each new file.
  5. Bare 'except:' is a bad idea.  Always catch an exception of a specific type.

As to what JP was saying in point 2: yes, he was saying that there was some redundant logic which should be eliminated and replaced with something common.  However, a new private function would be a more appropriate way to deal with this than a new public class; this functionality is only used internally in Twisted.

I'm going to reassign this to therve now, in the hopes that he will pick up where he left off, but if you would like to try again with a new patch, please feel free.

Thanks again,

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

The behavior described in the ticket summary isn't a regression, but the consequence this has for the new TLS implementation is one.  So marking this ticket as a regression overall, to reflect its blocker status on the 11.1 release.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32640]) Apply 5062-2.patch 

refs [#5062](https://github.com/twisted/twisted/issues/5062)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [32641]) and the other half of the patch

refs [#5062](https://github.com/twisted/twisted/issues/5062)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [32643]) Branching to 'wrapped-log-prefix-5062-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [32646]) Branching to 'wrapped-log-prefix-5062-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner|
|-|-|

I think I'm sorta happy-ish with how this is now.  Lots of code re-used from the previous branches, but some code shuffled around to avoid some duplication and to try to express the tests so they apply to more code without having to be rewritten for each transport/reactor.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/wrapped-log-prefix-5062-3)

Note I filed [#5278](https://github.com/twisted/twisted/issues/5278) for a related but separate issue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @exarkun|
|-|-|

Looks like a copy-paste error hit the docstrings for `UNIXTestsBuilder`.  I am pretty sure the **UNIX socket** tests builder is not building a **SSL endpoint** in its `serverEndpoint` and `clientEndpoint` methods.

Other than that it seems pretty good; addresses the issue, passes the tests, etc.  Please land after fixing that typo.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [32727]) We're not testing SSL over UNIX /yet/

refs [#5062](https://github.com/twisted/twisted/issues/5062)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `closed`|
|-|-|

(In [32728]) Merge wrapped-log-prefix-5062-3

Author: therve, retenodus, exarkun
Reviewer: zseil, exarkun, glyph
Fixes: [#5062](https://github.com/twisted/twisted/issues/5062)

Change transport implementations to use the protocol's ILoggingContext
implementation (when it is available) to construct the value for the
system key in the logging context, instead of always using the protocol
class's name directly.

This improves the behavior when using endpoints or `twisted.protocols.policies`
wrappers, since the wrapper protocol can include information about the wrapped
protocol in the log prefix.
